src/cets_join.erl

```diff
@@ -0,0 +1,78 @@
+%% @doc Cluster join logic.
+-module(cets_join).
+-export([join/4]).
+-include_lib("kernel/include/logger.hrl").
+
+%% Adds a node to a cluster.
+%% Writes from other nodes would wait for join completion.
+%% LockKey should be the same on all nodes.
+join(LockKey, Info, LocalPid, RemotePid) when is_pid(LocalPid), is_pid(RemotePid) ->
+    Info2 = Info#{local_pid => LocalPid,
+                  remote_pid => RemotePid, remote_node => node(RemotePid)},
+    F = fun() -> join1(LockKey, Info2, LocalPid, RemotePid) end,
+    cets_long:run_safely(Info2#{what => join_failed}, F).
+
+join1(LockKey, Info, LocalPid, RemotePid) ->
+    OtherPids = cets:other_pids(LocalPid),
+    case lists:member(RemotePid, OtherPids) of
+        true ->
+            {error, already_joined};
+        false ->
+                Start = erlang:system_time(millisecond),
+                F = fun() -> join_loop(LockKey, Info, LocalPid, RemotePid, Start) end,
+                cets_long:run(Info#{task => join}, F)
+    end.
+
+join_loop(LockKey, Info, LocalPid, RemotePid, Start) ->
+    %% Only one join at a time:
+    %% - for performance reasons, we don't want to cause too much load for active nodes
+    %% - to avoid deadlocks, because joining does gen_server calls
+    F = fun() ->
+        Diff = erlang:system_time(millisecond) - Start,
+        %% Getting the lock could take really long time in case nodes are
+        %% overloaded or joining is already in progress on another node
+        ?LOG_INFO(Info#{what => join_got_lock, after_time_ms => Diff}),
+        %% Do joining in a separate process to reduce GC
+        cets_long:run_spawn(Info, fun() -> join2(Info, LocalPid, RemotePid) end)
+        end,
+    LockRequest = {LockKey, self()},
+    %% Just lock all nodes, no magic here :)
+    Nodes = [node() | nodes()],
+    Retries = 1,
+    case global:trans(LockRequest, F, Nodes, Retries) of
+        aborted ->
+            ?LOG_ERROR(Info#{what => join_retry, reason => lock_aborted}),
+            join_loop(LockKey, Info, LocalPid, RemotePid, Start);
+        Result ->
+            Result
+    end.
+
+join2(_Info, LocalPid, RemotePid) ->
+    LocalOtherPids = cets:other_pids(LocalPid),
+    RemoteOtherPids = cets:other_pids(RemotePid),
+    LocPids = [LocalPid | LocalOtherPids],
+    RemPids = [RemotePid | RemoteOtherPids],
+    AllPids = LocPids ++ RemPids,
+    Paused = [{Pid, cets:pause(Pid)} || Pid <- AllPids],
+    %% Merges data from two partitions together.
+    %% Each entry in the table is allowed to be updated by the node that owns
+    %% the key only, so merging is easy.
+    try
+        cets:sync(LocalPid),
+        cets:sync(RemotePid),
+        {ok, LocalDump} = remote_or_local_dump(LocalPid),
+        {ok, RemoteDump} = remote_or_local_dump(RemotePid),
+        RemF = fun(Pid) -> cets:send_dump(Pid, LocPids, LocalDump) end,
```

Comment:
There would be a spike of memory usage (probably we need 2 times of the ETS table size).
For the real life ejabberd_sm tables - would be sent pretty fast.

Though, it is the fastest way to send. Also, could cause distributed connection processes to stuck for a second. 

I was thinking if we need pagination to be added - it could be done separately (separate PR), because it would be easier to review the main PR.